### PR TITLE
revise lesson 3 answer and add test

### PR DESCRIPTION
**Pull request by PatrickHo-134**

### CODE REVIEW

File: src/assignments/phat/lesson_4_answers.clj [hunk 1]
```diff
@@ -0,0 +1,161 @@
+(ns src.assignments.phat.lesson_4_answers)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:3]: Next time can you raise the PR just on this lesson's assignments?  I'm surprised to see two sets of answers in same PR.  Thanks.

File: src/assignments/phat/lesson_4_answers.clj [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+(def yippy-url "https://www.yippy.com/search?query=")
+
+(defn search
+  "return the first page of result from search engines"
+  [text]
+  (let [result (promise)]
+    (doseq [search-tools [bing-url yahoo-url]]
+      (future (if-let [results-page (slurp (str seacrh-tools text))]
+                (deliver result results-page))))
+    @result))
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:20]: Perfect!!!

File: src/assignments/phat/lesson_4_answers.clj [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+;; Update your function so it takes a second argument consisting of the search engines to use.
+(def search-engines {:bing bing-url
+                     :yahoo yahoo-url
+                     :yippy yippy-url})
+
+(defn search 
+  [text & tools]
+  (let [result (promise)]
+    (doseq [tool tools]
+      (assert (contains? search-engines tool)
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:36]: Nice touch!

File: src/assignments/phat/lesson_4_answers.clj [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+;;;;;;;;;;;;; 
+;; exercise 3
+;; Create a new function that takes a search term and search engines as arguments, and returns a vector of the URLs from the first page of search results from each search engine.
+(defn results-from-search [search-term & engines]
+  (let [results-page (search-on-engines search-term engines)
+        matches (re-seq #"href=\"([^\" ]*)\"" results)
+        uris (map second matches)
+        external-links (filter #(re-find #"http(?s)://" %) uris)]
+    external-links))
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:52]: sweet.  And good choice of names!

File: src/assignments/phat/lesson_4_answers.clj [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+(swap! person change-address "Melbourne")
+(swap! person grow-up)
+@person
+(swap! person grow-up)
+(swap! person grow-up)
+(swap! person change-address "Brisbane")
+@person
+(grow-up @person)
+person
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:100]: Yup!  You got it.

File: src/assignments/phat/lesson_4_answers.clj [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+;;;;;;;;;;;;;
+;; exercise 2
+
+(defn count-word-url
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:105]: Nope.  This function will block immediately.  Can you see why?

File: src/assignments/phat/lesson_4_answers.clj [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+(def doctor-strange (ref {:curr-hitpoint 20
+                          :max-hitpoint 40
+                          :inventory [:healing-potion :teleport :healing-potion]}))
+
+(defn remove-item
+  "remove an item from character`s inventory"
+  [character item]
+  (let [inventory (character :inventory)
+        item-pos (.indexOf inventory item)]
+    (vec (concat (subvec inventory 0 item-pos) (subvec inventory (inc item-pos))))))
```
Comment (alain-picard) [inline src/assignments/phat/lesson_4_answers.clj:139]: Try
(remove #{item} inventory)

Comment (PatrickHo-134) [reply src/assignments/phat/lesson_4_answers.clj:139]: I've tried it. Unfortunately, it removes all :healing-potion(s) in inventory if I have more than 1 and returns a sequence. So, in order to use remove, I have to change the representation of inventory somehow.

Best regards,

**Pull request merged**